### ISSUE TITLE
predict returns wrong number of predictions in R implementation with sparse data

### ISSUE THREAD

**Issue by JohnReid** [dmlc/xgboost#1503]
## Environment info
Operating System: Ubuntu 16.04

Compiler: gcc

Package used (python/R/jvm/C++): R

`xgboost` version used: 0.4.3

If you are using R package, please provide

1. The R `sessionInfo()`
```
R version 3.2.5 (2016-04-14)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.1 LTS

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] xgboost_0.4-3        Saturn_0.0.0.9000    dtplyr_0.0.1         reshape2_1.4.1       dplyr_0.5.0.9000     GenomicRanges_1.22.4 GenomeInfoDb_1.6.3   IRanges_2.4.8        S4Vectors_0.8.11     BiocGenerics_0.16.1  nvimcom_0.9-21       devtools_1.11.1      setwidth_1.0-4      
[14] colorout_1.1-1      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.6                       lattice_0.20-33                   Rsamtools_1.22.0                  Biostrings_2.38.4                 assertthat_0.1                    glmnet_2.0-5                      digest_0.6.9                      foreach_1.4.3                    
 [9] R6_2.1.2                          plyr_1.8.4                        futile.options_1.0.0              ggplot2_2.1.0                     zlibbioc_1.16.0                   lazyeval_0.2.0                    rstudioapi_0.5                    data.table_1.9.7                 
[17] Matrix_1.2-5                      rmarkdown_0.9.6                   BiocParallel_1.4.3                readr_0.2.2.9000                  stringr_1.0.0                     RCurl_1.95-4.8                    munsell_0.4.3                     rtracklayer_1.30.4               
[25] htmltools_0.3.5                   SummarizedExperiment_1.0.2        tibble_1.1                        roxygen2_5.0.1                    codetools_0.2-14                  XML_3.98-1.4                      BSgenome.Hsapiens.UCSC.hg19_1.4.0 crayon_1.3.2                     
[33] withr_1.0.1                       GenomicAlignments_1.6.3           bitops_1.0-6                      grid_3.2.5                        gtable_0.2.0                      DBI_0.4-1                         magrittr_1.5                      Romulus_1.0.1                    
[41] scales_0.4.0                      docopt_0.4.5                      stringi_1.1.1                     XVector_0.10.0                    ggthemes_3.0.3                    testthat_1.0.2                    futile.logger_1.4.1               PRROC_1.1                        
[49] functional_0.6                    lambda.r_1.1.7                    iterators_1.0.8                   tools_3.2.5                       BSgenome_1.38.0                   Biobase_2.30.0                    colorspace_1.2-6                  memoise_1.0.0                    
[57] prg_0.5.1                        

```

## Steps to reproduce
Extract model and data from attached zip file.
```
model <- xgb.load('model_file.xgb')
data <- readRDS('bug-features.rds')
predictions <- predict(model, data)
stopifnot(length(predictions) == nrow(data))
```

## What have you tried?
1. `predict(model, as.matrix(data))` works as expected
2. `predict(model, xgb.DMatrix(data))` fails in same way

[bug.zip](https://github.com/dmlc/xgboost/files/434497/bug.zip)

**Comment by JohnReid** [comment]
This actually seems to be a bug in `xgb.DMatrix` as the following code shows:
```
data.dm <- xgb.DMatrix(data)
stopifnot(nrow(data.dm) == nrow(data))
```

**Comment by JohnReid** [comment]
I followed this down to the C API and there does not seem any way to tell `XGDMatrixCreateFromCSC` what the desired dimensions of the matrix are. Is this by design? From a user of the R package's perspective the behaviour seems very odd that if final rows in your sparse matrix are all 0 then you get fewer predictions. For the moment I'm working around it with:

```
#'
#' Create a xgb.DMatrix from a dgCMatrix working around bug in
#' xgboost where zero rows at end of dgCMatrix are ignored.
#'
DMatrix.from.dgC <- function(dgC) {
  if (0 == dgC[nrow(dgC), 1]) {
    dgC[nrow(dgC), 1] <- .Machine$double.eps
  }
  xgb.DMatrix(dgC)
}

```

**Comment by hetong007** [comment]
The package `xgboost` seems to be outdated. Can you test it on the latest version 0.6-0?

**Comment by JohnReid** [comment]
0.4-3 is the latest version from CRAN. In any case there is the same behaviour on 0.6-0.

**Comment by hetong007** [comment]
The `0`s in the sparse matrix are treated as `NA` in xgboost. After `as.matrix(data)` they are numerical `0`s again. Usually we don't make prediction with an "entirely `NA`" data point.

**Comment by JohnReid** [comment]
Usually xgboost does make predictions with entirely `NA` data points as that's what it does when the rows are in the middle of the matrix, it just ignores those type of rows if they are at the end (but not at the beginning). That is inconsistent from a user's perspective. My data matrix has many rows just with zeros (or as you say `NA`s), some of them are used to predict but any at the end are not. Additionally I get a different result depending on the class of the object I pass into the predict argument. If I pass a dense R array I get all the predictions, if I pass a `dgCMatrix` or a `xgb.DMatrix` I'm missing the ones at the end. Even if this confusing behaviour is by design then it isn't easy to find in the documentation.

**Comment by hetong007** [comment]
It makes sense to me. You are welcome to open an PR for this issue. I will have time for the fix maybe in the next week.

**Comment by JohnReid** [comment]
I wanted to give it a go but then when I saw the change might be to the API of the underlying C library I was a bit wary as it would affect all the language bindings on top. If the only changes needed were in the R package I would have given it a go.

**Comment by hetong007** [comment]
I don't think a change in the R pacakge can do that. Your fix above is not the correct one since it changes an `NA` value into a small number, but sometimes the `NA` values are not equivalent to 0.

@tqchen Can we have your opinion on this? In short,  xgboost tends to ignore the all-`NA` rows at the end of the input sparse matrix but not the all-`NA` rows in the middle,  and we want to fix it.

**Comment by JohnReid** [comment]
My "fix" above isn't really a fix, just a way for me to get my code working when it is hard to predict how many predictions I will receive from `xgboost`. AFAICT the C API function `XGDMatrixCreateFromCSC` doesn't receive enough information to make more than an educated guess about the dimensions of the matrix it should return. This is perhaps why it drops trailing rows of zeros. It would need an additional argument to do know the dimensions. Having said that I'm not 100% sure about the sparse matrix representation being used so I could easily be wrong.

**Comment by khotilov** [comment]
In recent months, there were many reports because of this bug (which, in essence, is that automatic guessing of sparse data num_row or num_col for DMatrix is not working for CSC or CSR formats when there are empty rows or empty columns at the end), and there were some workarounds proposed and some ideas for how to possibly fix it. A proper fix would clearly require some API changes, so it would be very nice to get if not a solution but at least some direction for a feasible approach from the core developers, and to create a task for someone to pick up. My idea was to add an optional argument to XGDMatrixCreateFromCSR and to XGDMatrixCreateFromCSC which would allow to override the guessed num_col or num_row. But I am not fully confident right now whether that would be a reasonable way to do it, or whether any more serious redesign might be needed.

Here are some possible workaround options:
- From xgboost tree model's perspective, sparse data is equivalent to dense data with missing values. So you can convert sparse matrix to a dense one with with sparse values replaced with NA's (that can blow up your memory).
- For predictions with CSC, append a dummy row with at least one value to the end, then just discard the last predicted value.
- Append a dummy column with some constant value to the end of all your data (including the training set). That should generally work for both, CSC and CSR cases.

**Comment by hetong007** [comment]
Modifying the C API might affect the other interfaces, especially when adding a new argument. And it is a problem for all interfaces (including C++ command line). That's why I think @tqchen is the one to make decision.

The workaround options are all valid, but each of them brings some problems.
- The sparse to dense one, as stated, "blow up your memory".
- Adding a dummy row with one value to the end has no memory issue, but I would rather treat it as a dirty hack since we need to maintain the interface carefully to make sure it is consistent in every corner.
- Appending a dummy column does not affect the algorithm, it does not affect most of our implementation either. The problem is it reduces the sparsity and slightly slow down the training process. The advantage is the ease of implementation and maintainance of it. On the interface level, I personally think it is the one to go with.

**Comment by tqchen** [comment]
I agree with @khotilov on providing additional information to C API, we can add a new function called XGDMatrixCreateFromCSRV2 that add additional information, and move the API toward this before we removed the old ones completely.

PR will be more than welcomed

**Comment by tqchen** [comment]
https://github.com/dmlc/xgboost/issues/1583
